### PR TITLE
fix: add collect sonarqube mode by file

### LINKED ISSUE #6316: [Bug][sonarqube] Cannot get more than 10000 results through web API
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar issues.

### What happened

Cannot get more than 10000 results through web API

### What do you expect to happen

Can get more than 10000 results through web API

### How to reproduce

run sonarqube plugin.

### Anything else

_No response_

### Version

main

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### LINKED ISSUE #6493: [Feature][SonarQube] Get issues by files and line numbers to pass the result limit of 10,000 records
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar feature requirement.

### Use case

As a user who has around 200,000 issues per SonarQube project, I want to fetch all issues from SonarQube. The existing metrics in DevLake are inaccurate and different from those from SonarQube's UI.

Please make sure this method supports SonarQube versions later than v8.x.

### Description

Use file and line numbers to filter issues rather than projects. This can break up issues into smaller batches, the amount of which is smaller than 10,000.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by abeizn**
### Summary
fix: add collect sonarqube mode by file

### Does this close any open issues?
Closes #6493 

### Screenshots
image

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 1]
```diff
@@ -133,17 +145,43 @@ func CollectIssues(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+									Status:        status,
+									Type:          typ,
+									CreatedAfter:  createdAfter,
+									CreatedBefore: createdBefore,
+									FilePath:      value.Val,
+								})
+								logger.Info("split by dir for it's count:[%d] and val:[%s]", value.Count, value.Val)
+							} else {
+								// by file
+								for _, issue := range body.Issues {
```
Comment (leric) [inline backend/plugins/sonarqube/tasks/issues_collector.go:166]: why iterate issues here? might need another issues/search request to get files facets:
http://sonar_service/api/issues/search?componentKeys=be620d2f-b3fa-4c4c-9ddb-515a3810e908&directories=frontend%2Fapp%2Fpages%2Fproject-comparison-page&resolved=false&ps=1&facets=files

Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:166]: 1. xxxx?componentKeys=6f29ad17-7573-4042-85b5-3ea750c08fb1&directories=node/src/services/app&ps=1&facets=files     results show:
image
2. xxxx?componentKeys=6f29ad17-7573-4042-85b5-3ea750c08fb1:node/src/services/app result shows:
image

Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go]: updated

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 2]
```diff
@@ -133,17 +148,63 @@ func CollectIssues(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+									return 0, err
+								}
+								if len(body2.Facets) != 1 {
+									return 0, errors.Default.New(fmt.Sprintf("the facets count [%d] is not 1", len(body2.Facets)))
+								}
+								for _, value2 := range body2.Facets[0].Values {
+									if value2.Count > MAXISSUECOUNT {
+										logger.Warn(fmt.Errorf("the issue count [%d] exceeds the maximum page size", value2.Count), "")
+									} else {
+										iterator.Push(&SonarqubeIssueIteratorNode{
```
Comment (leric) [inline backend/plugins/sonarqube/tasks/issues_collector.go:189]: If the count exceed MAXISSUECOUNT, a warning is logged, we still need to push this node into the queue, or we should log an error.

Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:189]: Report an error? Or just ignore it？

Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go]: done

**Review by leric — APPROVED** [review]

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 3]
```diff
@@ -133,17 +148,62 @@ func CollectIssues(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+									Status:        status,
+									Type:          typ,
+									CreatedAfter:  createdAfter,
+									CreatedBefore: createdBefore,
+									FilePath:      value.Val,
+								})
+								logger.Info("split by dir, and it's issue count:[%d] and file path:[%s]", value.Count, value.Val)
+							} else {
+								// split it by dir when it's issue count > 10000
+								res2, err := data.ApiClient.Get("issues/search", url.Values{
```
Comment (d4x1) [inline backend/plugins/sonarqube/tasks/issues_collector.go:168]: res2 and body2  values2 should have  better names.

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 4]
```diff
@@ -133,17 +148,62 @@ func CollectIssues(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+									}
+									iterator.Push(&SonarqubeIssueIteratorNode{
+										Severity:      severity,
+										Status:        status,
+										Type:          typ,
+										CreatedAfter:  createdAfter,
+										CreatedBefore: createdBefore,
+										FilePath:      value2.Val,
+									})
+									logger.Info(fmt.Sprintf("split by fil, and it's issue count:[%d] and file path:[%s]", value2.Count, value2.Val))
```
Comment (d4x1) [inline backend/plugins/sonarqube/tasks/issues_collector.go:197]: no need to use fmt.Sprintf. logger.Info has a format parameter.

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 5]
```diff
@@ -239,3 +299,24 @@ func getTimeFromFormatTime(formatTime string) (*time.Time, errors.Error) {
 
 	return &t, nil
 }
+
+type SonarqubePageInfo struct {
+	Total  int `json:"total"`
+	P      int `json:"p"`
```
Comment (d4x1) [inline backend/plugins/sonarqube/tasks/issues_collector.go:305]: what's the meaning of p and ps?

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 6]
```diff
@@ -133,17 +148,62 @@ func CollectIssues(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+								err = helper.UnmarshalResponse(res2, body2)
+								if err != nil {
+									return 0, err
+								}
+								if len(body2.Facets) != 1 {
+									return 0, errors.Default.New(fmt.Sprintf("the facets count [%d] is not 1", len(body2.Facets)))
+								}
+								for _, value2 := range body2.Facets[0].Values {
+									if value2.Count > MAXISSUECOUNT {
+										logger.Warn(fmt.Errorf("the issue count [%d] exceeds the maximum page size", value2.Count), "")
```
Comment (d4x1) [inline backend/plugins/sonarqube/tasks/issues_collector.go:187]: no need to use fmt.Errorf.

File: backend/plugins/sonarqube/tasks/issues_collector.go [hunk 7]
```diff
@@ -133,17 +148,62 @@ func CollectIssues(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+									FilePath:      value.Val,
+								})
+								logger.Info("split by dir, and it's issue count:[%d] and file path:[%s]", value.Count, value.Val)
+							} else {
+								// split it by dir when it's issue count > 10000
+								res2, err := data.ApiClient.Get("issues/search", url.Values{
+									"componentKeys": {fmt.Sprintf("%v", data.Options.ProjectKey)},
+									"directories":   {value.Val},
+									"facets":        {"files"},
+									"ps":            {"1"},
```
Comment (d4x1) [inline backend/plugins/sonarqube/tasks/issues_collector.go:172]: Can this  work?  I think it should be  `"ps": {}string{"1"}`.

File: backend/plugins/sonarqube/tasks/issues_collector.go (hunk 5 above)
Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:305]: p: 1-based page number
ps: Page size. Must be greater than 0 and less or equal than 500

File: backend/plugins/sonarqube/tasks/issues_collector.go (hunk 6 above)
Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:187]: func (log.Logger).Warn(err error, format string, a ...interface{})
The first parameter is of error type, so it is required

File: backend/plugins/sonarqube/tasks/issues_collector.go (hunk 4 above)
Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:197]: done

File: backend/plugins/sonarqube/tasks/issues_collector.go (hunk 3 above)
Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:168]: done

File: backend/plugins/sonarqube/tasks/issues_collector.go (hunk 7 above)
Comment (abeizn) [reply backend/plugins/sonarqube/tasks/issues_collector.go:172]: yes,
"ps": {"1"} is correct
"ps": {}string{"1"} is invalid syntax

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**